Project selection dialog

ivanporty opened the pull request:
Final part of project selection change - dialog.
This looks like a large PR but it's actually not - at least 400 LOC are form XML/copyright headers/imports/visible for testing lines, and the rest cannot be split logically and should be straightforward to review.

You can review FilteredTextTableCellRenderer.java and ProjectListTableModelTest.java and their tests first - those are very small and independent, and continue with ProjectSelectionDialog.

Initial state - no logged in users:
![sign_in_screen](https://user-images.githubusercontent.com/11686100/33856504-532c2106-de96-11e7-9c6c-1bb1653d40fe.png)

Logged in users, no project selection:
![dialog](https://user-images.githubusercontent.com/11686100/33856331-cf7a1ad4-de95-11e7-9c69-803856bf18b9.png)

Filtered projects - both in our field and IDEA speed search on table (they are synced):
![dialog-filter](https://user-images.githubusercontent.com/11686100/33856452-31226eda-de96-11e7-83c4-f76bd863c027.png)

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectListTableModel.java:
```diff
@@ -0,0 +1,75 @@
+        return projectList.get(row).getProjectId();
+    }
+
+    return "";
```
if you are using a switch, why not just use a `default` case?

ivanporty replied:
done.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectListTableModel.java:
```diff
@@ -0,0 +1,75 @@
+        return GctBundle.getString("project.selector.project.list.project.id.column");
+    }
+
+    return "";
```
same comment

ivanporty replied:
done.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectorSignInPanel.java:
```diff
@@ -0,0 +1,56 @@
+class ProjectSelectorSignInPanel extends BaseGoogleLoginUi {
+
+  ProjectSelectorSignInPanel() {
+    super(GctBundle.message("select.project.signin"));
```
Didn't catch this last time time, but perhaps a better name here? Something along the lines of `cloud.project.selector.signin.message`

ivanporty replied:
good idea, changing here and adding cloud.project.selector to other strings instead of just project.selector

etanshaul commented:
Instead of "Filter Name/ID", how about "Search by name or ID"?

ivanporty commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java:
```diff
@@ -33,6 +95,334 @@
+
+  @VisibleForTesting
+  void loadUsersAndProjects() {
+    Collection<CredentialedUser> credentialedUsers =
```
It does select active user by default, see line 152, but will keep the order of users in the dropdown as it is. I can change it to have the active user on top.

etanshaul replied:
oops, didn't catch that. I think its fine as is.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java:
```diff
@@ -33,6 +95,334 @@
+
+  @VisibleForTesting
+  void setCloudProject(CloudProject cloudProject) {
+    this.cloudProject = cloudProject;
```
this also feels like an opportunity for refactoring. setting this as a member variable feels off to me.

ivanporty replied:
We can remove this variable and just use it once to set the state of the account combobox and project list selection - I think we won't need to keep it after that. WDYT?

ivanporty replied:
done.

ivanporty commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java:
```diff
@@ -33,6 +95,334 @@
+  @VisibleForTesting
+  void setCloudProject(CloudProject cloudProject) {
+    this.cloudProject = cloudProject;
+    updateProjectAccountInformation();
```
In loadUsersAndProjects it loads all accounts and their projects and selects active (default) account - i.e. prepares default state of the dialog.
In updateProjectAccountInformation it changes the state of the dialog to match that of the passed CloudProject. How about setProjectSelectionTo(CloudProject)?

ivanporty replied:
Removed updateProjectAccountInformation() altogether.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java:
```diff
@@ -33,6 +95,334 @@
-    return cloudProject;
+    // create dialog wrapper for the form once, create services and subscribe to login changes once.
+    if (dialogWrapper == null) {
+      dialogWrapper = new ProjectSelectionDialogWrapper();
```
I'm trying to wrap my head around this. Why didn't you want this class to be a `DialogWrapper` itself?

etanshaul replied:
in other words, why not just extend from DialogWrapper? As is it is, it is very tightly coupled to DialogWrapper anyway, so I'm not seeing the advantage. I could be wrong though.

ivanporty replied:
I tried it first and inheriting from `DialogWrapper` brings a lot of internal UI implementation with it - I had to spy/mock/override a lot of internal methods, and it started 'leaking projects' because I used it - something was not disposing properly in mock test mode. It's much easier and cleaner to test and edit as a simple form, but we have to have this wrapper class to pass dialog wrapper buttons and additional actions. It is not very friendly to composition instead of inheritance.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java:
```diff
@@ -33,6 +103,346 @@
+      if (loggedInUser.isPresent()) {
+        selectedProjectsByAccount.put(loggedInUser.get(), cloudProject.projectName());
+        accountComboBox.setSelectedItem(loggedInUser.get());
+        updateProjectList(loggedInUser.get());
```
why is this needed here if the project list is already loaded (and cached) in #loadUsersAndProjects ?

etanshaul replied:
oh..I think i see..thinking about this

etanshaul replied:
actually, nm. If the cached list is already loaded, why do we need to load it again for this specific user?

ivanporty replied:
It won't be loaded since it's cached, I sort of added it for increased readability so reader doesn't need to imply project will be selected because of the combox box event. We can remove it safely.

etanshaul replied:
ah ok. lets remove it then. it had the opposite effect on me. made me confused. :)

ivanporty replied:
done.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java:
```diff
@@ -33,6 +103,346 @@
+                          this::loadUsersAndProjects));
+    }
+
+    loadUsersAndProjects();
```
I know we can go back and forth forever on naming, but just thinking after looking again..how about:

loadUsersAndProjects -> loadAllProjects
setCloudProject -> setActiveProject (I think the "cloud" part is implied enough)

etanshaul replied:
actually not setActiveProject, but setSelectedProject

ivanporty replied:
sounds good, and I'll change the getter name to reflect that idea as well - i.e. `getSelectedProject`

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java:
```diff
@@ -33,6 +103,348 @@
+      accountComboBox.removeAllItems();
+      for (CredentialedUser user : credentialedUsers) {
+        accountComboBox.addItem(user);
+        updateProjectList(user);
```
This is what I was thinking, lmkwyt

- `loadAllProjects` is responsible for loading and caching the projects for all users. Thats is; it doesnt update the UI.
- `setSelectedProject` then updates the dropdown with the current CloudProject, and in turn, causes the table model to update (pulling the list of projects from the cache).
- the listener on the account combobox updates the table model like in the above bullet

Its similar, but I think this will separate concerns a bit more cleanly. Sorry if I am contradicting previous things I've said with these comments. The motivation is I think that updateProjectList is a bit too complicated and doing too much (loading projects, updating the UI..).

ivanporty replied:
sounds reasonable, I'll do a simple refactor of this idea, trying to clean up `updateProjectList' a bit more

etanshaul commented:
Can be done in a separate PR later, but it might be nice to have some loading indicator when the projects are being refreshed (or when a new account is selected and the project list is loading).

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java:
```diff
@@ -16,15 +16,85 @@
+  private Map<CredentialedUser, ListenableFuture<List<Project>>> runningProjectLoaderJobs =
+      Maps.newHashMap();
+  // last project name selection made by user for an account.
+  private Map<CredentialedUser, String> selectedProjectsByAccount = Maps.newHashMap();
```
what do you think about combining selecteProjectsByAccount and cachedProjectList into a single mutlimap?

etanshaul replied:
specifically, guava's multimap

ivanporty replied:
you mean selected project is marked somehow in project list, i.e. we have Multimap<User,Project> and selected project is one of them?

etanshaul replied:
Never mind, it may have been a bad idea.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java:
```diff
@@ -33,6 +103,348 @@
+      Optional<CredentialedUser> loggedInUser =
+          Services.getLoginService().getLoggedInUser(cloudProject.googleUsername());
+      if (loggedInUser.isPresent()) {
+        selectedProjectsByAccount.put(loggedInUser.get(), cloudProject.projectName());
```
Cloud project name is not unique; project ID is. I just want to make sure the logic here and elsewhere is aware of this. In other words, if this is actually using the cloud project name here, then I think this example usage will have problems (since it may not select the correct project).

ivanporty replied:
so in `AppEngineDeploymentConfiguration`, the field is `cloudProjectName`, is it actually name or ID?

etanshaul replied:
The main thing is that whenever we are keying on it, we must use the ID. But if we are displaying it (like in the new Cloud Project widget - [name] / [email]) we use the name.

ivanporty replied:
sounds good, then let's add  `projectId` to keep the actual ID in `CloudProject` too

ivanporty commented:
Cleaned up `updateProjectList` a bit more, renamed to better understand what it does.
Also, added loading indicator - using small neat indicator in top right corner from IDEA table.
PTAL at changes.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java:
```diff
@@ -33,6 +104,363 @@
+  }
+
+  private void loadProjectList(CredentialedUser user) {
+    // check if cached, if not then check if already loading, and finally start loading if not.
```
could you promote top level comments like this that apply to the whole method to actually be javadoc method comments? I know some of these cases are for private methods, but I still think it makes sense.

ivanporty replied:
sure, great idea.

etanshaul commented:
I dunno if its a bug or not, but the loader doesn't consistently show up. I just clicked the refresh button a bunch of times and sometimes it appeared and sometimes not.

Also, maybe move the loader to be beside the refresh button? It took me several seconds to find it even when it did appear.

etanshaul commented:
or even in the middle of table (if thats possible). wherever you think makes more sense.

etanshaul commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java:
```diff
@@ -33,6 +104,363 @@
+  }
+
+  @VisibleForTesting
+  void loadAllProjects() {
```
thanks for the refactoring. much easier to read now

ivanporty replied:
yeah it is much easier to read, good catch!

etanshaul commented:
possible bug:

1) sign in to multiple accounts
2) select the second account in the list
3) click refresh

-> I'd expect the same account's projects to be refreshed. Instead it sent me back to the other account.

ivanporty commented:
Added clearly visible centered overlay progress bar, preserved account selection when refresh button is hit.
